✨ enable capbk to read secrets

alexeldeib opened the pull request:
**What this PR does / why we need it**:
Example of API types for https://docs.google.com/document/d/1R_E3uu8Uvfe4pOcTrfQYUolJCK8mUS58Kq4A8zmqTr8/edit?usp=sharing

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

alexeldeib commented on bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go:
```diff
@@ -175,6 +175,98 @@ type File struct {
+//
+// The contents of the target ConfigMap's Data field will be presented
+// as files using the keys in the Data field as the file names.
+type ConfigMapFileSource struct {
```
ConfigMapFileSource and SecretFileSource  are identical, wonder if we can simplify that with a typed object reference?

fabriziopandini commented on bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go:
```diff
@@ -47,7 +47,7 @@ type KubeadmConfigSpec struct {
 	// Files specifies extra files to be passed to user_data upon creation.
 	// +optional
-	Files []File `json:"files,omitempty"`
+	Files []FileSource `json:"files,omitempty"`
```
If we are going for this approach this way this is a breaking change, it should way for v1alpha4...

alexeldeib replied:
fixed for backcompat

fabriziopandini commented on bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go:
```diff
@@ -175,6 +175,98 @@ type File struct {
+
+	// ConfigMap represents a configMap that should populate this volume
+	// +optional
+	ConfigMap *ConfigMapFileSource `json:"configMap,omitempty"`
```
Is there a use case for ConfigMaps? 
I'm debated betweek ok, if we have secrets then also ConfigMaps are ok, and let's keep this simple...

alexeldeib replied:
removed until someone asks for it

fabriziopandini commented on bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go:
```diff
@@ -175,6 +175,98 @@ type File struct {
+
+	// Secret represents a secret that should populate this file.
+	// +optional
+	Secret *SecretFileSource `json:"secret,omitempty"`
```
I'm a little bit confused by this, because SecretFileSource is a specialization of  FileSource, but it seems that it allows acting as a source for many files (like a mount)

alexeldeib replied:
Correct, based on my reading of https://github.com/kubernetes-sigs/cluster-api/pull/1860#discussion_r355594849 and similar, this is desired behavior.

The thinking was that a user would specify the source one time, but could project multiple files (keys in the data of the secret). If you specify raw content, the file source can only produce one output.

alexeldeib commented:
@fabriziopandini I probably need to align this with the mounts proposal, I didn't consider the overlap between the two.

The approach with FileSources was meant to directly mirror what everyone seemed okay with on https://github.com/kubernetes-sigs/cluster-api/pull/1860, specifically https://github.com/kubernetes-sigs/cluster-api/pull/1860#discussion_r358479010 with the adjustments as in https://github.com/kubernetes-sigs/cluster-api/pull/1860#discussion_r359423322. I'm fine with both Files and FileSources for back-compat, just wanted to get some discussion going.

alexeldeib commented:
@fabriziopandini could you re-review? I've split out files/fileSources for backcompat and added some unit tests with examples.

I also removed ConfigMaps until we hear a strong case for it -- my thinking was some sort of frequent multi-file re-use, but we can keep the API minimal until someone asks for more.

I don't think this should have any impact on the etcd/disk mount proposal

alexeldeib commented:
@CecileRobertMichon @devigned 

This now works

```yaml
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
kind: KubeadmConfigTemplate
metadata:
  name: ${CLUSTER_NAME}-md-0
spec:
  template:
    spec:
      fileSources:
      - secret: 
          secretName: azure-secret
          items:
          - path: /etc/kubernetes/azure.json
            permissions: "0644"
            owner: "root:root"
            key: "azure.json"
      joinConfiguration:
        nodeRegistration:
          kubeletExtraArgs:
            cloud-config: /etc/kubernetes/azure.json
            cloud-provider: azure
          name: '{{ ds.meta_data["local_hostname"] }}'
---
apiVersion: v1
kind: Secret
metadata:
  name: azure-secret
data:
  "azure.json": |
    {
      "cloud": "AzurePublicCloud",
      "tenantId": "XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
      "subscriptionId": "XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
      "aadClientId": "XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
      "aadClientSecret": "XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
      "resourceGroup": "capz-cluster",
      "securityGroupName": "capz-cluster-node-nsg",
      "location": "southcentralus",
      "vmType": "vmss",
      "vnetName": "capz-cluster-vnet",
      "vnetResourceGroup": "capz-cluster",
      "subnetName": "capz-cluster-node-subnet",
      "routeTableName": "capz-cluster-node-routetable",
      "loadBalancerSku": "standard",
      "maximumLoadBalancerRuleCount": 250,
      "useManagedIdentityExtension": false,
      "useInstanceMetadata": true
    }
```

vincepri commented:
/assign

vincepri commented on bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go:
```diff
@@ -49,6 +49,14 @@ type KubeadmConfigSpec struct {
+	// in plain text may be stored in external sources and resolved at
+	// runtime.
+	// +optional
+	FileSources []FileSource `json:"fileSources,omitempty"`
```
Should this be `FilesFrom`?

alexeldeib replied:
That's what was discussed in the other PR but I felt the alignment with volumes with configmap and secret source made sense, no reason to diverge

vincepri commented on bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go:
```diff
@@ -175,6 +183,71 @@ type File struct {
+	Secret *SecretFileSource `json:"secret,omitempty"`
+}
+
+type FileSourceOptions struct {
+	// DefaultPermissions specifies optional mode bits to use on created files by default.
+	// Must be a value between 0 and 0777.
+	// +optional
+	DefaultPermissions string `json:"defaultPermissions,omitempty"`
+
+	// DefaultOwner specifies the ownership of the file, e.g. "root:root".
+	// +optional
+	DefaultOwner string `json:"defaultOwner,omitempty"`
+
+	// DefaultEncoding specifies the encoding of the file contents.
+	// +optional
+	DefaultEncoding Encoding `json:"defaultEncoding,omitempty"`
+}
```
If these are all optional, is there a default for these defaults?

alexeldeib replied:
cloud-init defaults, will add to doc comments.

vincepri commented on bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go:
```diff
@@ -175,6 +183,71 @@ type File struct {
+	// Name of the secret in the KubeadmBootstrapConfig's namespace to use.
+	SecretName string `json:"secretName"`
+
+	// If unspecified, each key-value pair in the Data field of the referenced
+	// Secret will be projected as a file whose name is the
+	// key and content is the value. If specified, the listed keys will be
+	// projected into the specified paths, and unlisted keys will not be
+	// present. If a key is specified which is not present in the Secret,
+	// the volume setup will error.
+	// +optional
+	Items []KeyToPath `json:"items,omitempty"`
```
This effectively gives a way to users to do the same thing in multiple ways. Should we instead find a single solution, at least for now, that by contract has 1 secret for each file you'd want to inject?

A few benefits:
- API would be simpler (no need for defaults, and defaults for those defaults).
- Secrets wouldn't be able to grow too much (etcd storage limitations).
- Clear separation between each secret.
- No need to look at keys, we'd force each secret to have a single `value` secret.

alexeldeib replied:
quick note before the rest of my response, I discovered yesterday that secret data keys don't map well to file paths (can't use slashes like in config maps). So we definitely need some form of metadata to map the data keys into file paths.

I think 1 value per source is okay, I do think there's a case for reuse that will come up eventually. I don't have a firm scenario right now for Azure, as azure.json is 1 file per cluster, and 1 per secret currently. 

After trying this, pattern *might* emerge (at least for Azure) for users to have 1 secret per cluster with all the sensitive data fully spec'd out, and they don't have to jump around with references. This would happen kind of naturally since we'd already need 1 secret per cluster due to the uniqueness of cloud provider file.

I think future-proofing has some benefits, given the resulting API is pretty usable for the single-file, single-key case while expanding the alternate API would be less flexible (breakage around key/item mapping. I don't really like the top level expansion of defaultPermissions, defaultEncoding, etc. -- maybe we should force that to be fully qualified?

any other thoughts? @ncdc since you were on the original discussion too?

alexeldeib commented:
@vincepri @fabianofranz any additional thoughts on this one?

fabriziopandini commented:
> @fabriziopandini I probably need to align this with the mounts proposal, I didn't consider the overlap between the two.

If I'm not wrong this is not addressed yet, but might be @vincepri prefers the current approach 
(@vincepri FYI. My original comment was about the idea that secrets can be a different type of mount instead of a file source, so we have a nice similarity with volume mounts in Pods and potentially we can leverage on some core types instead of introducing our types)

alexeldeib commented:
I read the etcd proposal and discussed with @CecileRobertMichon offline. I left a comment in https://github.com/kubernetes-sigs/cluster-api/pull/3038#issuecomment-629052737

Neither of us really felt it made sense to merge these. Looking at the field configuration, we take disjoint inputs. There's probably an edge case for something like "I want a file from a secret mounted onto a data disk after that disk gets mounted", but that's pretty far from a reason to merge disks + files in one mount-like structure IMO. 

If we merged disks and file sources, we'd have one field for raw files, and another field that produces a mix of disks and files, accepting totally different inputs?

I think the notion of similarity makes sense for pods and object/field refs, but in this case I think the disks mounts are different. The disk is IaaS we can't have beforehand but need to know about at runtime. The file sources are he opposite -- we know about them up front and just want to plumb the values through securely. 

So I think keeping files + disks as peers makes sense, or possibly merging to the non-backwards compatible version where merging files + file sources, which share similar inputs, makes sense. I don't see a case for how one would align the implementations of file + disk mounts -- if you have a concrete example given the difference in the field values, I'd love to see it. I might just have tunnel vision here.

vincepri commented:
Briefly chatted today with @alexeldeib and @CecileRobertMichon, we'll trim down the addition to the following changes:

- Have a 1:1 relationship between a Secret and a File
- Make the `File.Content` field optional and omitempty.
- Add `File.ContentFrom` that allows users to reference a Secret (name + key).
- Add some validation to make sure we have either `File.Content` or `File.ContentFrom` populated.

@alexeldeib will follow-up in a new PR when ready.

alexeldeib commented:
/close
